numpy/random/mtrand.pyi

```diff
@@ -0,0 +1,588 @@
+    def __repr__(self) -> str: ...
+    def __str__(self) -> str: ...
+    def __getstate__(self) -> Dict[str, Any]: ...
+    def __setstate__(self, state: Dict[str, Any]) -> None: ...
+    def __reduce__(self) -> Tuple[Callable[[str], BitGenerator], Tuple[str], Dict[str, Any]]: ...
+    def seed(self, seed: Optional[_ArrayLikeFloat_co] = ...) -> None: ...
+    @overload
+    def get_state(self, legacy: Literal[False] = ...) -> Dict[str, Any]: ...
+    @overload
+    def get_state(
+        self, legacy: Literal[True] = ...
+    ) -> Union[Dict[str, Any], Tuple[str, ndarray[Any, dtype[uint32]], int, int, float]]: ...
+    def set_state(
+        self, state: Union[Dict[str, Any], Tuple[str, ndarray[Any, dtype[uint32]], int, int, float]]
+    ) -> None: ...
+    @overload
+    def random_sample(self, size: None = ...) -> float: ...  # type: ignore[misc]
+    @overload
+    def random_sample(self, size: _ShapeLike = ...) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def random(self, size: None = ...) -> float: ...  # type: ignore[misc]
+    @overload
+    def random(self, size: _ShapeLike = ...) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def beta(self, a: float, b: float, size: None = ...) -> float: ...  # type: ignore[misc]
+    @overload
+    def beta(
+        self, a: _ArrayLikeFloat_co, b: _ArrayLikeFloat_co, size: Optional[_ShapeLike] = ...
+    ) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def exponential(self, scale: float = ..., size: None = ...) -> float: ...  # type: ignore[misc]
+    @overload
+    def exponential(
+        self, scale: _ArrayLikeFloat_co = ..., size: Optional[_ShapeLike] = ...
+    ) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def standard_exponential(self, size: None = ...) -> float: ...  # type: ignore[misc]
+    @overload
+    def standard_exponential(self, size: _ShapeLike = ...) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def tomaxint(self, size: None = ...) -> int: ...  # type: ignore[misc]
+    @overload
+    def tomaxint(self, size: _ShapeLike = ...) -> ndarray[Any, dtype[int_]]: ...
+    @overload
+    def randint(  # type: ignore[misc]
+        self,
+        low: int,
+        high: Optional[int] = ...,
+    ) -> int: ...
+    @overload
+    def randint(  # type: ignore[misc]
+        self,
+        low: int,
+        high: Optional[int] = ...,
+        size: None = ...,
+        dtype: _DTypeLikeBool = ...,
+        endpoint: bool = ...,
+    ) -> bool: ...
+    @overload
+    def randint(  # type: ignore[misc]
+        self,
+        low: int,
+        high: Optional[int] = ...,
+        size: None = ...,
+        dtype: _DTypeLikeBool = ...,
+    ) -> bool: ...
+    @overload
+    def randint(  # type: ignore[misc]
+        self,
+        low: int,
+        high: Optional[int] = ...,
+        size: None = ...,
+        dtype: Union[_DTypeLikeInt, _DTypeLikeUInt] = ...,
+    ) -> int: ...
```

Comment:
Uhm, so what's going on here with these three completely different signatures?

Reply:
I couldn't get it to handle the default values properly.  I read around on the mypy  github about complex defaults and this is what I came up with.  Probably no minimal.

Reply:
I've been able to clean this up a lot.

Reply:
I just realized I selected the wrong lines; it's really these first two overloads that I don't understand (_e.g._ what's this `endpoint` parameter doing here all of a sudden?).
``` python
    @overload
    def randint(  # type: ignore[misc]
        self,
        low: int,
        high: Optional[int] = ...,
    ) -> int: ...
    @overload
    def randint(  # type: ignore[misc]
        self,
        low: int,
        high: Optional[int] = ...,
        size: None = ...,
        dtype: _DTypeLikeBool = ...,
        endpoint: bool = ...,
    ) -> bool: ...
```

Reply:
I need the default path to come first since this produces an `int`, then the more specific `bool` path, then the full `int` path.

I followed what I learned from 

https://github.com/python/mypy/issues/5486

when going with this strategy. I think the problem is that a bool is an int but not vice versa. If I omit the short def, and then have the `int` def first, then the scalar `bool` is reported as an `int`, which isn't totally wrong.  This strategy allowed me to get the most specific type information.